src/neighborsearch/neighborhood.jl

```diff
@@ -63,8 +76,69 @@ function search(xₒ::AbstractVector, method::NeighborhoodSearch{O,N,T};
         push!(neighbors, ind)
       end
     end
+    method.restrictions && (neighbors = filterneighs(neighbors, method, xₒ))
     neighbors
   else
+    method.restrictions && (inds = filterneighs(inds, method, xₒ))
     inds
   end
 end
+
+function filterneighs(inds::AbstractVector, method::NeighborhoodSearch{O,N,T},
+  xₒ::AbstractVector) where {O,N<:BallNeighborhood,T}
+
+  object   = method.object
+  neigh    = method.neigh
+  rotmat   = neigh.metric isa Mahalanobis ? neigh.metric.qmat : I
+  ometric  = method.ordermetric
+  maxk     = method.maxneighs
+  maxoct   = method.maxperoct
+  maxkey   = method.maxperkey
+  usek     = 0 < maxk < length(inds)
+  useoct   = maxoct > 0
+  usekey   = maxkey[2] > 0
+
+  # get distances and give priority to closest neighbors according to ordermetric
+  dists   = colwise(ometric, xₒ, coordinates(object,inds))
+  sortids = sortperm(dists)
+  inds    = inds[sortids]
+
+  # key and octant constraints
+  keys   = usekey ? unique(view(values(object), inds, maxkey[1])) : nothing
+  dkeys  = usekey ? Dict(zip(keys, 1:length(keys)))               : nothing
+  ctkeys = usekey ? zeros(Int, length(keys))                      : nothing
+  octs   = useoct ? zeros(Int, 8)                                 : nothing
```

Comment:
We need to refactor this code to make it more easy to follow. I see that there is some logic to initialize some auxiliary variables in the filtering that comes later. Am I correct that there two disjoint filtering operations happening in the loop? If that is the case, I would create two separate sections in the function body, one for the filtering with `usekey` and another section for the filtering with `useoct`. The second section could use the result of the first section for performance reasons and clarity.